apps/server/src/modules/authentication/strategy/ldap.strategy.ts

```diff
@@ -0,0 +1,58 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+import { Injectable, UnauthorizedException } from '@nestjs/common';
+import { PassportStrategy } from '@nestjs/passport';
+import { ICurrentUser } from '@shared/domain';
+import { CurrentUserMapper } from '@shared/domain/mapper/current-user.mapper';
+import { SchoolRepo, SystemRepo, UserRepo } from '@shared/repo';
+import { Strategy } from 'passport-custom';
+import { AuthenticationService } from '../services/authentication.service';
+import { LdapService } from '../services/ldap.service';
+
+export type RequestBody = { systemId?: string; username?: string; password?: string; schoolId?: string };
+
+@Injectable()
+export class LdapStrategy extends PassportStrategy(Strategy, 'ldap') {
+	constructor(
+		private readonly systemRepo: SystemRepo,
+		private readonly schoolRepo: SchoolRepo,
+		private readonly ldapService: LdapService,
+		private readonly authenticationService: AuthenticationService,
+		private readonly userRepo: UserRepo
+	) {
+		super();
+	}
+
+	async validate(request: { body: RequestBody }): Promise<ICurrentUser> {
+		const { systemId, schoolId } = request.body;
+		let { username, password } = request.body;
+		if (!systemId || !username || !password || !schoolId) {
```

Comment:
Please put all if throw to private checkXXX methods.

Reply:
Won't do. I get TS errors like "Argument of type 'string | undefined' is not assignable to parameter of type 'string'.", if I do the checks in extra methods. I would need to create type guards and define extra classes like `SchoolWithMandatoryExternalId`.

I don't think it improves readability.

Reply:
this.checkIfNeededValuesAvaible(request.body); 
this.checkExternalId(school);
this.checkUserId(account);
this.checkLdapDn(user);

On which place it need define something?
How to check it is a other level as do the check. We want to do the login in this method. For this we need to exectute the checks, but not the implementation of the checks it self.

Reply:
If I externalize the check like this
´´´
this.checkExternalId(school);
const userNameWithSchool = `${school.externalId}/${username}`
```
I get a typescript error in the second line, saying that `school.externalId` might be undefined.
I can only do this, if the return type of the internal function is a type guard. It looks something like this:

```
checkExternalId(school: School): arg is SchoolWithExternalId {
    return school.externalId !== undefined;
}

validate(...)
    ...
    if(!this.checkExternalId(school) {
        throw new UnauthorizedException();
    }
}
```
So I need to keep the if statement and define a type `SchoolWithExternalId`